cpp/serve/engine_actions/batch_verify.cc

```diff
@@ -0,0 +1,403 @@
+  }
+
+  /*! \brief Check if the drafts can be verified under conditions. */
+  bool CanVerify(EngineState estate, int num_verify_req, int total_draft_length,
+                 int num_required_pages, int num_available_pages) {
+    int num_running_requests = estate->running_queue.size();
+    ICHECK_LE(num_running_requests, kv_cache_config_->max_num_sequence);
+
+    // No exceeding of the maximum allowed requests that can
+    // run simultaneously.
+    if (num_running_requests + num_verify_req > kv_cache_config_->max_num_sequence) {
+      return false;
+    }
+
+    // NOTE: The conditions are heuristic and can be revised.
+    // Cond 1: total input length <= max allowed single sequence length.
+    // Cond 2: at least one verify can be performed.
+    // Cond 3: number of total tokens does not exceed the limit
+    int new_batch_size = num_running_requests + num_verify_req;
+    return total_draft_length <= max_single_sequence_length_ &&
+           num_required_pages <= num_available_pages &&
+           estate->stats.current_total_seq_len + total_draft_length <=
+               kv_cache_config_->max_total_sequence_length;
+  }
+
+  /*!
+   * \brief Decide whether to run verify for the draft of each request.
+   * \param estate The engine state.
+   * \return The drafts to verify, together with their respective
+   * state and input length.
+   */
+  std::tuple<Array<Request>, Array<RequestState>, std::vector<int>, int> GetDraftsToVerify(
+      EngineState estate) {
+    // - Try to verify pending requests.
+    std::vector<Request> verify_requests;
+    std::vector<RequestState> rstates;
+    std::vector<int> draft_lengths;
+    int total_draft_length = 0;
+    int total_required_pages = 0;
+    int num_available_pages = models_[verify_model_id_]->GetNumAvailablePages();
+
+    int req_id = 1;
+    for (; req_id <= static_cast<int>(estate->running_queue.size()); ++req_id) {
+      Request request = estate->running_queue[req_id - 1];
+      RequestState rstate = estate->GetRequestState(request);
+      int draft_length = rstate->mstates[draft_model_id_]->GetDraftLength();
+      int num_require_pages =
+          (draft_length + kv_cache_config_->page_size - 1) / kv_cache_config_->page_size;
+      total_draft_length += draft_length;
+      total_required_pages += num_require_pages;
+      if (CanVerify(estate, req_id, total_draft_length, total_required_pages,
+                    num_available_pages)) {
+        verify_requests.push_back(request);
+        rstates.push_back(rstate);
+        draft_lengths.push_back(draft_length);
+      } else {
+        total_draft_length -= draft_length;
+        total_required_pages -= num_require_pages;
+        break;
+      }
+    }
+    // preempt all the remaining requests
+    // TODO: can we remove requests that are in the middle positions?
+    while (req_id <= static_cast<int>(estate->running_queue.size())) {
+      PreemptLastRunningRequest(estate);
+      req_id += 1;
+    }
```

Comment:
Theoretically yes, but doing that may hurt the latency or the request that comes first I think?